### ISSUE TITLE
Models with time-varying boundaries

### ISSUE THREAD

**Issue by DominiqueMakowski** [itsdfish/SequentialSamplingModels.jl#41]
Was just wondering about the possibility of implementing models with time-varying boundaries (or collapsing boundaries). That said, it might make sense to not have that as a priority, given some of their critiques.

- https://www.frontiersin.org/articles/10.3389/fpsyg.2014.01364/full (implementation)
- https://www.jneurosci.org/content/35/6/2476 (a critique of these models)
- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5450920/ (another critique)

**Comment by itsdfish** [comment]
@DominiqueMakowski, I agree with your assessment. There is some controversy regarding the validity of time varying boundaries. I'm definitely not opposed to adding support at some point. 

I think this is a good opportunity to create a working wishlist of models we could add. Let me know if there is anything else you want to add. 

# Model Wishlist

| model                 	| assigned 	| priority 	| references 	| finished 	|
|-----------------------	|----------	|----------	|------------	|----------	|
| time varying boundary 	| NA       	| low      	|     1,2,3       	| No       	|
| mlba                  	| itsdfish 	| high     	|        4    	| No       	|


# References
1. [Zhang, S., Lee, M. D., Vandekerckhove, J., Maris, G., & Wagenmakers, E. J. (2014). Time-varying boundaries for diffusion models of decision making and response time. Frontiers in psychology, 5, 1364.](https://www.frontiersin.org/articles/10.3389/fpsyg.2014.01364/full) 
2. [Hawkins, G. E., Forstmann, B. U., Wagenmakers, E. J., Ratcliff, R., & Brown, S. D. (2015). Revisiting the evidence for collapsing boundaries and urgency signals in perceptual decision-making. Journal of Neuroscience, 35(6), 2476-2484.](https://www.jneurosci.org/content/35/6/2476) 
3. [Voskuilen, C., Ratcliff, R., & Smith, P. L. (2016). Comparing fixed and collapsing boundary versions of the diffusion model. Journal of mathematical psychology, 73, 59-79.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5450920/) 
4. [Trueblood, J. S., Brown, S. D., & Heathcote, A. (2014). The multiattribute linear ballistic accumulator model of context effects in multialternative choice. Psychological review, 121(2), 179.](http://www.psy.vanderbilt.edu/jdmlab/papers/Trueblood_PsychReview2014.pdf)

**Comment by DominiqueMakowski** [comment]
[rtdists](https://github.com/rtdists/rtdists/)seems to implement LBA with varying drift SDs. Is this a feature that we could add?

**Comment by itsdfish** [comment]
Yeah. I pushed a new [release](https://github.com/JuliaRegistries/General/pull/88657) which uses a vector for the drift SDs. The default is a vector of ones.

**Comment by DominiqueMakowski** [comment]
I suspect the same could be done for LNR, but I don't know if that's common

**Comment by itsdfish** [comment]
I was wondering the same thing myself. I need to verify its identifiable when the SDs differ. I believe it is.

**Comment by DominiqueMakowski** [comment]
Also, to add to the wishlist above:

multiple-threshold log-normal race (MTLNR): [Reynolds et al. (2020)](https://www.sciencedirect.com/science/article/abs/pii/S0022249620300171) :)

**Comment by DominiqueMakowski** [comment]
- [ ] Time Varying Extension of the Shifted Wald Model: [Howard et al., in press](https://osf.io/preprints/psyarxiv/4wq9r/)
  - [R code](https://osf.io/dz7v2)
- [ ] Generalised exponential-Gaussian distribution (GEG models): [Marmolejo-Ramos et al., 2022](https://link.springer.com/article/10.1007/s11571-022-09813-2)

**Comment by kiante-fernandez** [comment]
I am also interested in the time varying shifted wald. Let me read that paper this weekend and give a shot at the translation.

**Comment by itsdfish** [comment]
Sounds good. I will take a look at the  Generalised exponential-Gaussian model.

**Comment by itsdfish** [comment]
I decided against adding the generalized Ex-Gaussian because it does not appear to be theoretically motivated. I added the Ex-Gaussian instead.  The psychological interpretation of the Ex-Gaussian is dubious too, but I added it because it was simple and it has a historical role in RT modeling.

**Comment by DominiqueMakowski** [comment]
> but I added it because it was simple and it has a historical role in RT modelling

Yes, totally agree, very cool! 

> because it does not appear to be theoretically motivated

Out of curiosity, can you elaborate a bit?

**Comment by itsdfish** [comment]
I did not go through the papers in fine detail, but my tentative impression is that the validity of the ex-gaussian distribution has not been strongly established, and the new shape parameter of the generalized ex-gaussian model has no psychological interpretation. If the $\alpha$ changes, what does that mean in terms of cognitive processes or mechanisms? Probably nothing. 

I recommend checking out the paper below. The authors show through simulation and experimentation that the ex-Gaussian parameters do not change selectively with experimental manipulations (or varying DDM parameters). In many cases, all parameters respond to the manipulations, which seems to imply they do not map onto cognitive processes in a clear fashion if at all.  

_Matzke, D., & Wagenmakers, E. J. (2009). Psychological interpretation of the ex-Gaussian and shifted Wald parameters: A diffusion model analysis. Psychonomic bulletin & review, 16, 798-817._

**Comment by kiante-fernandez** [comment]
> * [ ]  Time Varying Shifted Wald Model: [Howard et al., in press](https://osf.io/preprints/psyarxiv/4wq9r/)
>   
>   * [R code](https://osf.io/dz7v2)

After reviewing the specific paper you mentioned, @DominiqueMakowski, it appears to primarily focus on exploiting a particular aspect of the Wald distribution's relationship with another distribution (e.g., ex-Gaussian). This approach is employed to facilitate on-the-fly computation of the posterior through conjugacy. Such trial-varying estimation permits variation in parameters across different trials. Initially, however, I had the impression that you were interested in exploring dynamics within trials, such as fluctuations in drift in the ADDM model, for example. If our focus remains on the former aspect, I believe only minor adjustments (if any) are necessary for the current Wald implementation. Subsequently, it would simply be a matter of demonstrating the procedure as an example of how trial-varying estimation could be implemented in Julia. However, this seems more like an illustrative application rather than the development of an entirely new model for the package. What are your guys thoughts on this?

**Comment by itsdfish** [comment]
> on-the-fly computation of the posterior through conjugacy

I only gave the paper a cursory read, but arrived at a similar conclusion and wanted to wait for your evaluation. There is no need to create a new model type because the Wald is applied to a moving window of data. In my opinion, the biggest innovation was deriving a conjugate prior, which is difficult except in very simple cases. I think it would be sufficient to add an example to the parameter estimation examples in the docs showing how to use the conjugate prior without the moving window of data. The only thing that was a bit unclear to me was how non-decision time was treated.

**Comment by DominiqueMakowski** [comment]
👍

**Comment by DominiqueMakowski** [comment]
Sorry to just add stuff on this thread while being totally useless regarding their assessment and implementation ^^ 

Unless I'm mistaken, most models can currently already be used for one-choice paradigms (e.g., LBA or LNR with one accumulator; + all the one-choice specific models like exgaussian and Wald). Most but the DDM, currently (?).

[This paper](https://www.pnas.org/doi/10.1073/pnas.1100483108) apparently shows that it's doable, but I didn't find many details regarding the implementation. Is that something that we could support?

**Comment by itsdfish** [comment]
@DominiqueMakowski, I think these ideas are all worthy of discussion and are not necessarily useless.  Conjugate priors are quite convenient and have practical utility.  So I think that is worth mentioning it in the parameter estimation section. 

I think the single choice DDM is something that we could support. Note that it does not have a known PDF and the drift rate mean and SD are not identifiable. One alternative that I would suggest is the WaldMixture, which is similar except it assumes non-decision time is a constant rather than a random variable. The WaldMixture has a known PDF and the mean and SD drift rates are identifiable. 

If you are interested in false starts (premature responses), the paper below extends the 1 choice DDM to such situations. It has the same issues, but it models the pre-stimulus decision process. 


_Walsh, M. M., Gunzelmann, G., & Van Dongen, H. P. (2017). Computational cognitive modeling of the temporal dynamics of fatigue from sleep loss. Psychonomic bulletin & review, 24, 1785-1807._

**Comment by DominiqueMakowski** [comment]
- [hybrid racing-diffusion ex-Gaussian(RDEX) stop-signal model](https://www.researchgate.net/publication/377324993_A_hybrid_approach_to_dynamic_cognitive_psychometrics_Dynamic_cognitive_psychometrics): though it doesn't look like a new model per se but rather just using 2 exGaussian ones if I understand correctly

However, in the future, it might be useful to have more examples in the documentation where we show how to implement these published SSM applications with Julia

**Comment by itsdfish** [comment]
@DominiqueMakowski, thanks for the suggestions. I'll try to find time to read the paper for the RDEX in more detail. After a cursory read, it looks like it is applied to the Go, No-Go task where the racing diffusion is used for go responses and the ex-Gaussian is used for no go responses. I'm not sure exactly why this mixture is used, but that will probably become clearer once I read the paper more thoroughly. 

I can add a page to the document that explains how to extend the API for new models. I think that is a good idea.

**Comment by kiante-fernandez** [comment]
Circling back to the time varying idea, I was exploring simulations for a model that captures the concept of certain sources of information coming "online" at different time points. This has been discussed in several applied cases where the drift considered multiple attributes, weighting their importance and changing their latency relative to one another (in the two-attribute case). These are either called starting time DDM (stDDM) or multi-attribute, time-dependent DDM (mtDDM). I put together a simulator for this idea since it is rampant in the literature. Here is a link to the PR:

https://github.com/itsdfish/SequentialSamplingModels.jl/pull/59


Here are some citations of use cases:

- Amasino, D.R., Sullivan, N.J., Kranton, R.E., et al. Amount and time exert independent influences on intertemporal choice. Nat Hum Behav 3, 383–392 (2019). https://doi.org/10.1038/s41562-019-0537-2
- Barakchian, Z., Beharelle, A.R., & Hare, T.A. Healthy decisions in the cued-attribute food choice paradigm have high test-retest reliability. Sci Rep, (2021). https://doi.org/10.1038/s41598-021-91933-6
- Chen, H.Y., Lombardi, G., Li, S.C., et al. Older adults process the probability of winning sooner but weigh it less during lottery decisions. Sci Rep, (2022). https://doi.org/10.1038/s41598-022-15432-y
- Lombardi, G., & Hare, T. Piecewise constant averaging methods allow for fast and accurate hierarchical Bayesian estimation of drift diffusion models with time-varying evidence accumulation rates. PsyArXiv, (2021). https://doi.org/10.31234/osf.io/5azyx
- Sullivan, N.J., Huettel, S.A. Healthful choices depend on the latency and rate of information accumulation. Nat Hum Behav 5, 1698–1706 (2021). https://doi.org/10.1038/s41562-021-01154-0

**Comment by DominiqueMakowski** [comment]
-**Nonparametric-bound Drift-Diffusion Model (npb-DDM)**: Seems similar to the time-collapsing-boundaries models -https://www.researchgate.net/publication/377858197_Incorporation_of_a_cost_of_deliberation_time_in_perceptual_decision_making)

![image](https://github.com/itsdfish/SequentialSamplingModels.jl/assets/8875533/65402faa-bfd2-4db0-a2ea-05de5266f0ad)

**Comment by DominiqueMakowski** [comment]
- [Wang 2024](https://link.springer.com/article/10.3758/s13423-024-02520-5): Suggests that the Ornstein-Uhlenbeck Model (OUM) model, a variant of the DDM, might be able to capture more neurally plausible mechanisms than the DDM.
- Small helpsheet on [priors](https://www.gingjehli.com/single-post/choosing-effective-samplers-and-setting-priors) in DDMs

**Comment by itsdfish** [comment]
The LCA is conceptually similar. I am not sure how or whether they might differ.  A while back I started an implemention of the OUM,  but I was unsure whether it was correct. I will look through the article to see whether I can get some clues.

**Comment by DominiqueMakowski** [comment]
Just stumbled on the **Autocorrelated Bayesian Sampler (ABS)** model, which appears like a "one model to rule them all" kind of model 

- [Zhu, J. Q., Sundh, J., Spicer, J., Chater, N., & Sanborn, A. N. (2023). The autocorrelated Bayesian sampler: A rational process for probability judgments, estimates, confidence intervals, choices, confidence judgments, and response times. Psychological review.](https://www.researchgate.net/publication/371414835_The_autocorrelated_Bayesian_sampler_A_rational_process_for_probability_judgments_estimates_confidence_intervals_choices_confidence_judgments_and_response_times)

Quite an interesting paper, reframing the role of time in cognitive models.

**Comment by itsdfish** [comment]
Thanks for sharing. Its been a while since I read that article. I agree that they are interesting ideas. One thing I remember is that some of the model fits were not very good, which is difficult to interpret without understanding to what extend core and auxilary assumptions are driving the predictions.  In either case, I do appreciate the effort to broaden the scope of theory.

**Comment by DominiqueMakowski** [comment]
**Sequential sampling without comparison to boundary through model-free reinforcement learning**

> Although evidence integration to the boundary model has successfully explained a wide range of behavioral and neural data in decision making under uncertainty, how animals learn and optimize the boundary remains unresolved. Here, we propose a model free reinforcement learning algorithm for perceptual decisions under uncertainty that dispenses entirely with the concepts of decision boundary and evidence accumulation. Our model learns whether to commit to a decision given the available evidence or continue sampling information at a cost


- https://arxiv.org/pdf/2408.06080

**Comment by itsdfish** [comment]
I think it would be reasonable to add a model with an urgancy signal, especially if the physiological evidence is compelling. One reason I have not added such a model is that it was unclear to me whether it was better supported than standard alternatives. These different mechanisms are challenging to distinguish with choice-RT data alone, and might not be possible once auxiliary assumptions are removed (see Jones & Dzhafarov, 2014).
